lib/ascii_binder/engine.rb

```diff
@@ -0,0 +1,872 @@
+              subtopic_list << {
+                :path => "../#{topic_group['Dir']}/#{topic['Dir']}/#{subtopic['File']}.html",
+                :name => subtopic['Name'],
+                :id   => subtopic['ID'],
+              }
+            end
+            topic_list << { :name => topic['Name'], :id => topic['ID'], :topics => subtopic_list }
+          end
+        end
+        navigation << { :name => topic_group['Name'], :id => topic_group['ID'], :topics => topic_list }
+      end
+      navigation
+    end
+
+    def asciidoctor_page_attrs(more_attrs=[])
+      [
+        'source-highlighter=coderay',
+        'coderay-css=style',
+        'linkcss!',
+        'icons=font',
+        'idprefix=',
+        'idseparator=-',
+        'sectanchors',
+        'data-uri',
+      ].concat(more_attrs)
+    end
+
+    def generate_docs(branch_group,build_distro,single_page)
+      # First, test to see if the docs repo has any commits. If the user has just
+      # run `asciibinder create`, there will be no commits to work from, yet.
+      if local_branches.empty?
+        raise "Before you can build the docs, you need at least one commit in your docs repo."
+      end
+
+      single_page_dir  = []
+      single_page_file = nil
+      if not single_page.nil?
+        single_page_dir  = single_page.split(':')[0].split('/')
+        single_page_file = single_page.split(':')[1]
+        puts "Rebuilding '#{single_page_dir.join('/')}/#{single_page_file}' on branch '#{working_branch}'."
+      end
+
+      if not build_distro == ''
+        if not distro_map.include_distro_key?(build_distro)
+          exit
+        else
+          puts "Building only the #{distro_map.get_distro(build_distro).name} distribution."
+        end
+      elsif single_page.nil?
+        puts "Building all distributions."
+      end
+
+      # First, notify the user of missing local branches
+      missing_branches = []
+      distro_map.distro_branches(build_distro).sort.each do |dbranch|
+        next if local_branches.include?(dbranch)
+        missing_branches << dbranch
+      end
+      if missing_branches.length > 0 and single_page.nil?
+        puts "\nNOTE: The following branches do not exist in your local git repo:"
+        missing_branches.each do |mbranch|
+          puts "- #{mbranch}"
+        end
+        puts "The build will proceed but these branches will not be generated."
+      end
+
+      # Generate all distros for all branches in the indicated branch group
+      branch_group_branches[branch_group].each do |local_branch|
+        # Skip known missing branches; this will only come up for the :publish branch group
+        next if missing_branches.include?(local_branch)
+
+        # Single-page regen only occurs for the working branch
+        if not local_branch == working_branch
+          if single_page.nil?
+            # Checkout the branch
+            puts "\nCHANGING TO BRANCH '#{local_branch}'"
+            git_checkout(local_branch)
```

Comment:
I think this is the reason for #80 

Branch 1 contains a ditaa or other diagram that gets rendered by asciidoctor-diagram during build. Then we have committed a version of the resulting image file into the branch. For this branch processing runs and re-generates the file but with a different checksum. Now when you try to move to branch 2 which also contains a version of this image file git will complain that these differ and that changes must be stashed. 

My problem here is that in jenkins I can not catch this error and so the build fails. Can we do a stash on every change of the branch to avoid this problem?

In line 722 when changing the working branch you do a stash all.

Not 100% sure but this looks like what's happening.